### PR TITLE
[Mamba POC] Handle UpdateModifier.UPDATE_DEPS correctly (all graph, not just leaves)

### PR DESCRIPTION
**Pull request by jaimergp**
This tries to fix `test_update_deps_flag_present`. Mamba code only handled the leave nodes I think. In this case we take conda's approach and do a pre-solve to obtain the full chain.

I've also reworked how the solver tasks are precomputed here. This might affect other tests too. Keep an eye on regressions.

### CODE REVIEW

File: conda/core/solve.py [hunk 1]
```diff
@@ -1324,28 +1345,65 @@ def _configure_solver_for_remove(self, state):
...
+            force_remove=NULL,
+            force_reinstall=NULL):
         """
         Reimplement the logic found in super()._add_specs(), but simplified.
         """
+        from mamba import mamba_api as api
+        task_and_specs = defaultdict(set)
         # A. Explicitly requested packages via CLI (either update or install)
-        specs_to_add = list(self.specs_to_add)
+        task = (api.SOLVER_UPDATE if self._command == "update" else api.SOLVER_INSTALL)
```
Comment (jezdez) [inline conda/core/solve.py:1361]: No need for extra parentheses.
```suggestion
        task = api.SOLVER_UPDATE if self._command == "update" else api.SOLVER_INSTALL
```

File: conda/core/solve.py [hunk 2]
```diff
@@ -1324,28 +1345,65 @@ def _configure_solver_for_remove(self, state):
...
+            specs_map = {name: MatchSpec(name) for name in update_names}
+
+            # Remove pinned_specs and any python spec (due to major-minor pinning business rule).
+            # Add in the original specs_to_add on top.
+            for spec in self._pinned_specs():
+                specs_map.pop(spec.name, None)
+            if "python" in specs_map:
+                python_rec = next((p for p in state["installed_pkgs"] if p.name == "python"), None)
+                if python_rec is not None:
+                    py_ver = ".".join(python_rec.version.split(".")[:2]) + ".*"
```
Comment (jezdez) [inline conda/core/solve.py:1401]: This seems a bit brittle, is there a chance that `python_rec.version` could be an empty string or nonsensible version?

File: conda/core/solve.py [hunk 3]
```diff
@@ -1363,9 +1421,9 @@ def _pin_python(self, state, update_modifier=NULL):
             return f"python {major_minor_version}.*"
 
     def _pinned_specs(self):
-        pinned_specs = get_pinned_specs(context.target_prefix)
+        pinned_specs = get_pinned_specs(self.prefix)
```
Comment (jezdez) [inline conda/core/solve.py:1424]: Could you elaborate why this change?

File: conda/core/solve.py [hunk 4]
```diff
@@ -1324,28 +1345,65 @@ def _configure_solver_for_remove(self, state):
...
+                deps_modifier=deps_modifier,
+                ignore_pinned=ignore_pinned,
+                force_remove=force_remove,
+                force_reinstall=force_reinstall)
+            graph = PrefixGraph(solved_pkgs)
+            update_names = set()
+            for spec in self.specs_to_add:
+                node = graph.get_node_by_name(spec.name)
+                update_names.update(ancest_rec.name for ancest_rec in graph.all_ancestors(node))
+            specs_map = {name: MatchSpec(name) for name in update_names}
```
Comment (jezdez) [inline conda/core/solve.py:1392]: Could we move this into a matching PrefixGraph method?

Comment (jezdez) [reply conda/core/solve.py:1392]: Actually including the removal of pinned specs below?

**Review by jezdez — APPROVED** [review]
Just a few nits and questions :)

**Comment by jaimergp** [comment]
Oops, sorry Jannis, I wasn't finished with this yet. I'll mark it as draft.

**Comment by jezdez** [comment]
No problem ;)

**Comment by jaimergp** [comment]
Phew, it shows that we are entering tricky territory now. Let's review this for now @jezdez

File: conda/core/solve.py (hunk 3 above)
Comment (jaimergp) [reply conda/core/solve.py:1454]: Some tests do pass the prefix explicitly to the class (which defaults to `context.target_prefix` if `prefix` is not passed explicitly).

File: conda/core/solve.py [hunk 5]
```diff
@@ -1324,28 +1347,93 @@ def _configure_solver_for_remove(self, state):
...
+            for spec in specs_to_add_plus_deps:
+                node = graph.get_node_by_name(spec.name)
+                update_names.update(ancest_rec.name for ancest_rec in graph.all_ancestors(node))
+            specs_map = {name: MatchSpec(name) for name in update_names}
+
+            # Remove pinned_specs and any python spec (due to major-minor pinning business rule).
+            for spec in self._pinned_specs():
+                specs_map.pop(spec.name, None)
+            if "python" in specs_map:
+                python = next((p for p in state["installed_pkgs"] if p.name == "python"), None)
```
Comment (jezdez) [inline conda/core/solve.py:1415]: Let's add a TODO item here that we may want to abstract this into a utility API for other cases in which we want to restrict this spec map.

**Review by jezdez — APPROVED** [review]
Great, some nits.

**Pull request merged**